xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -187,6 +206,65 @@ void watchConfigData(String hostName, int port, ConfigWatcher watcher) {
     adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
   }
 
+  @Override
+  void watchClusterData(String clusterName, ClusterWatcher watcher) {
+    checkNotNull(watcher, "watcher");
+    boolean needRequest = false;
+    if (!clusterWatchers.containsKey(clusterName)) {
+      logger.log(Level.FINE, "Start watching cluster {0}", clusterName);
+      needRequest = true;
+      clusterWatchers.put(clusterName, new HashSet<ClusterWatcher>());
+    }
+    Set<ClusterWatcher> watchers = clusterWatchers.get(clusterName);
+    if (watchers.contains(watcher)) {
+      logger.log(Level.WARNING, "Watcher {0} already registered", watcher);
+    }
+    watchers.add(watcher);
+    // If local cache contains cluster information to be watched, notify the watcher immediately.
+    if (clusterNamesToClusterUpdates.containsKey(clusterName)) {
+      watcher.onClusterChanged(clusterNamesToClusterUpdates.get(clusterName));
+    }
+    if (rpcRetryTimer != null) {
+      // Currently in retry backoff.
+      return;
+    }
+    if (needRequest) {
+      if (adsStream == null) {
```

Comment:
I cannot comment on `startRpcStream` because it's old code. But I think there are sync issues here:

- you are calling it if `adsStream == null`
- you are throwing an exception on line 347 if `adsStream != null` (`checkState` call)

And  `adsStream.start()` on line 351 is potentially long-running operation so you can have a situation where in another thread you have adsStream assigned and waiting on `adsStream.start()`. In this thread you just skip and go to line 235 `adsStream.sendXdsRequest` which is going to fail because `requestWriter` is either null or old.

I think you have a race condition between code on lines 231-236 (and similar code elsewhere) and `startRpcStream` . You should protect using a lock.

Reply:
I think I had suggested in a previous comment, you can have something like this:
```
  private synchronized void startRpcStreamIfNeeded() {
    if (adsStream == null) {
       startRpcStreamInternal();
    }
  }

  private void startRpcStreamInternal() {
    AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub =
        AggregatedDiscoveryServiceGrpc.newStub(channel);
    adsStream = new AdsStream(stub);
    adsStream.start();
    stopwatch.reset().start();
  }

  private void startRpcStream() {
    checkState(adsStream == null, "Previous adsStream has not been cleared yet");
    startRpcStreamInternal();
}
```
You can probably have just one function, need to think

Reply:
Everything in this class runs in the same `SynchronizationContext` as that caller uses to invoke methods of this class. So there isn't threading issues, every field is set sequentially.

Reply:
ok